File: Python/Enfriamiento Cuántico (versión heurística inspirada).py
import math
import random
import time

# Función para calcular la distancia entre dos puntos
def distancia(p1, p2):
    return math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

# Función para calcular el largo de un recorrido
def largo_recorrido(ruta, puntos):
    total = 0
    for i in range(len(ruta) - 1):
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]])
    return total

# Enfriamiento cuántico (versión heurística inspirada)
def enfriamiento_cuantico(puntos, n_iteraciones=1000, temperatura_inicial=1000, temperatura_final=0.01, alpha=0.995):
    n = len(puntos)
    mejor_ruta = list(range(n))
    mejor_largo = largo_recorrido(mejor_ruta, puntos)

    ruta = mejor_ruta[:]
    temperatura = temperatura_inicial

    while temperatura > temperatura_final:
        for _ in range(n_iteraciones):
            i, j = random.sample(range(n), 2)  # Elegir dos índices aleatorios
            ruta[i], ruta[j] = ruta[j], ruta[i]  # Intercambiar las posiciones

            largo = largo_recorrido(ruta, puntos)

            if largo < mejor_largo or math.exp((mejor_largo - largo) / temperatura) > random.random():
                mejor_ruta = ruta[:]
                mejor_largo = largo
            else:
                # Si no aceptamos, revertimos el cambio
                ruta[i], ruta[j] = ruta[j], ruta[i]

        temperatura *= alpha

    return mejor_ruta, mejor_largo

# Ejemplo de uso
if __name__ == '__main__':
    # Generación de puntos aleatorios
    puntos = [(random.randint(0, 100), random.randint(0, 100)) for _ in range(10)]

    start_time = time.time()
    mejor_ruta, mejor_largo = enfriamiento_cuantico(puntos)
    end_time = time.time()

    # Imprimir el resultado
    print("Mejor ruta:", mejor_ruta)
    print("Distancia total:", mejor_largo)
    print("Tiempo de ejecución:", end_time - start_time, "segundos")

File: c++/Enfriamiento Cuántico (versión heurística inspirada).cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <limits>

using namespace std;

// Función de distancia
double distancia(pair<int, int> p1, pair<int, int> p2) {
    return sqrt(pow(p1.first - p2.first, 2) + pow(p1.second - p2.second, 2));
}

// Función para calcular el largo de un recorrido
double largo_recorrido(vector<int>& ruta, vector<pair<int, int>>& puntos) {
    double total = 0;
    for (int i = 0; i < ruta.size() - 1; ++i) {
        total += distancia(puntos[ruta[i]], puntos[ruta[i + 1]]);
    }
    return total;
}

// Enfriamiento cuántico (versión heurística inspirada)
pair<vector<int>, double> enfriamiento_cuantico(vector<pair<int, int>>& puntos, int n_iteraciones = 1000, double temperatura_inicial = 1000, double temperatura_final = 0.01, double alpha = 0.995) {
    int n = puntos.size();
    vector<int> mejor_ruta(n);
    for (int i = 0; i < n; ++i) mejor_ruta[i] = i;
    double mejor_largo = largo_recorrido(mejor_ruta, puntos);

    vector<int> ruta = mejor_ruta;
    double temperatura = temperatura_inicial;

    while (temperatura > temperatura_final) {
        for (int iter = 0; iter < n_iteraciones; ++iter) {
            // Realizamos un intercambio de dos ciudades aleatorias
            int i = rand() % n;
            int j = rand() % n;
            swap(ruta[i], ruta[j]);

            // Calculamos el largo de la nueva ruta
            double largo = largo_recorrido(ruta, puntos);

            // Si la nueva ruta es mejor, la aceptamos, si no, con cierta probabilidad
            if (largo < mejor_largo || exp((mejor_largo - largo) / temperatura) > (rand() / double(RAND_MAX))) {
                mejor_ruta = ruta;
                mejor_largo = largo;
            } else {
                // Si no la aceptamos, revertimos el intercambio
                swap(ruta[i], ruta[j]);
            }
        }

        // Reducir la temperatura
        temperatura *= alpha;
    }

    return make_pair(mejor_ruta, mejor_largo);
}

// Ejemplo de uso
int main() {
    srand(time(0));

    // Generación de puntos aleatorios (por ejemplo, 10 puntos)
    vector<pair<int, int>> puntos;
    for (int i = 0; i < 10; ++i) {
        puntos.push_back({rand() % 100, rand() % 100});
    }

    clock_t start_time = clock();

    // Llamar al algoritmo de Enfriamiento Cuántico
    pair<vector<int>, double> resultado = enfriamiento_cuantico(puntos);

    vector<int> mejor_ruta = resultado.first;
    double mejor_largo = resultado.second;

    clock_t end_time = clock();

    // Imprimir el resultado
    cout << "Mejor ruta:";
    for (int i : mejor_ruta) {
        cout << " " << i;
    }
    cout << "\nDistancia total: " << mejor_largo << endl;
    cout << "Tiempo de ejecución: " << double(end_time - start_time) / CLOCKS_PER_SEC << " segundos\n";
}
